server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1189,6 +1208,26 @@ private static class MetadataDiff implements Diff<Metadata> {
             new DiffableUtils.DiffableValueReader<>(ReservedStateMetadata::readFrom, ReservedStateMetadata::readDiffFrom);
 
         MetadataDiff(StreamInput in) throws IOException {
+            if (in.getVersion().onOrAfter(NOOP_METADATA_DIFF_VERSION)) {
```

Comment:
Do we have a unit tests that checks serialization/deserialization of the MetadataDiff? DO we need one for an empty case?

Reply:
Turns out we don't, it's massively covered by existing tests but it's nice to have a test that makes sure we get instance equality for the empty diff to lock down this optimization. I added one now.